### DISCUSSION TITLE
Does 43-ccd.py work for CCD analytic gradients?

### DISCUSSION THREAD

**Discussion by 1234zou** [pyscf/pyscf#2681] [Q&A] [answered]
The script examples/cc/43-ccd.py offers a simple solution to calculate the CCD energy using CCSD code. If `mc.Gradients()` is appended to calculate the CCD analytic gradients, will the results be correct?

I've performed such a calculation using different quantum chemistry packages, and I find that CCD analytic gradients of PySCF are somewhat different from those of other packages. This is my input script
[h2o_rhf.zip](https://github.com/pyscf/pyscf/files/13114865/h2o_rhf.zip)

Here are the last few lines in the script
```
mc = cc.CCSD(mf)
mc.verbose = 4
mc.frozen = 0
old_update_amps = mc.update_amps
def update_amps(t1, t2, eris):
  t1, t2 = old_update_amps(t1, t2, eris)
  return t1*0, t2
mc.update_amps = update_amps
mc.kernel()

from pyscf import grad
mcg = mc.Gradients()
mcg.max_memory = 4000 # MB
mcg.kernel()
```

Here are the CCD/cc-pVTZ analytic gradients from 3 packages (no frozen core):
(1) Gaussian
```
  0.00000000E+00   0.00000000E+00 -4.16132400E-03
  0.00000000E+00  -1.17245720E-02   2.08066200E-03
  0.00000000E+00   1.17245720E-02   2.08066200E-03
```

(2) CFOUR
```
  0.00000000E+00   1.10000000E-14  -4.16106176E-03
  0.00000000E+00  -1.17247788E-02   2.08053088E-03
  0.00000000E+00   1.17247788E-02   2.08053088E-03
```

(3) PySCF
```
  0.00000000E+00   0.00000000E+00 -4.63005710E-03
  0.00000000E+00  -1.14598343E-02   2.31502860E-03
  0.00000000E+00   1.14598343E-02   2.31502860E-03
```

Maybe there is some code to be modified regarding solving the lambda equation? Thanks for any help or suggestion.

**Comment by tberkel** [comment]
Just to eliminate any other possible sources of disagreement, can you
confirm that if you use CCSD, you get better agreement with gradients
between codes?

**Comment by 1234zou** [comment]
@tberkel Thank you for your reply. Yes. Here are CCSD/cc-pVTZ analytic gradients from 3 packages (no frozen core, the same geometry as that of CCD):
(1) Gaussian
```
  0.00000000E+00   0.00000000E+00  -5.28802200E-03
  0.00000000E+00  -1.11307350E-02    2.64401100E-03
  0.00000000E+00   1.11307350E-02   2.64401100E-03
```

(2) CFOUR
```
  0.00000000E+00   6.00000000E-15  -5.28775389E-03
  0.00000000E+00  -1.11309438E-02   2.64387695E-03
  0.00000000E+00   1.11309438E-02   2.64387695E-03
```

(3) PySCF
```
  0.00000000E+00   0.00000000E+00  -5.28818420E-03
  0.00000000E+00  -1.11307931E-02   2.64409210E-03
  0.00000000E+00   1.11307931E-02   2.64409210E-03
```
One can see the better agreement. As for the electronic energies, even better agreement can be observed, so I did not shown any energy data.

**Comment by verena-neufeld** [comment]
I have a hunch that it has to do with how the lambda equations are solved, https://github.com/pyscf/pyscf/blob/169257b680df5f508283cde404f4f373f686cbfe/pyscf/grad/ccsd.py#L438 . I haven't thought about how they should be defined though. Maybe l1 should be set to zero?

**Comment by verena-neufeld** [comment, marked as answer]
Still not 100% sure, but since we only have even excitations due to T in CCD, we should probably only have even deexcitations with lambda, i.e. I think probably l1 = 0 in CCD.
